### ISSUE TITLE
(chore) Only do certain checks for documentation changes (github workflow)

### ISSUE THREAD

**Issue by weedySeaDragon** [mermaid-js/mermaid#3898]
### Description

There are specific github workflows and checks unique to documentation (but not code):
 - checking links

Likewise, there are workflows and checks specific to code and not documentation:
- unit tests
- end to end tests (cypress e2e)
- test static analysis

We should modify the github workflow(s) so that _only_ the workflows & checks required for the context (documentation or code) are run.  If a PR includes changes to both documentation and code, then all relevant workflows/checks should be run.

Recently the link checker has been failing because of some issue communicating with a documentation link destination. This is happening _even if only code was changed._ It means that a PR fails and a contributor must be involved and review and resolve/override/re-run the failing check.  Obviously this means it takes more time for the PR to be completed and means that more people have to be invovled.

Only running the necessary workflows/checks will reduce PR submission friction for contributors:  it makes the process faster and smoother, which makes it more likely that more contributions will be made because it's less onerous.

### Steps to reproduce

Only change a wee bit of documentation.
Submit a new PR.
Note all of the code-specific checks/workflows/actions that have to be run.

**Comment by weedySeaDragon** [comment]
Here's a workflow action that could be used: [Changed Files](https://github.com/marketplace/actions/changed-files)

**Comment by sidharthv96** [comment]
Running link checker only on documentation changes is an excellent idea. 
This might introduce a small issue, though. Now link checker is running in all PRs, and it might be keeping the cache warm.
If we switch it to only documentation PRs, it's almost guaranteed to fail, with GitHub links throwing 429s on the first run. 

We can experiment with the cache duration and maybe run the workflow using cron.

**Comment by sidharthv96** [comment]
Can we use the native feature https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-paths ?


I had experimented with changed-files for the auto build docs pr here https://github.com/mermaid-js/mermaid/pull/3547/commits/9c5c85d34ac1777a93b8e80b15720a5fccbeb57c. It works pretty well, but is an extra dependency.

**Comment by weedySeaDragon** [comment]
Sure -- the native feature would work.
What are the limits of how many dependencies are "too many"?  It's a classic "build vs. buy" trade-off: someone else is responsible for maintaining that code vs. more code to maintain in this project.  
Just trying to get a handle on what the priorities / loads/ costs are for this project.

Not a big deal to me - I think the bigger payoff will be making documentation easier and faster to submit, and likewise code submission won't get hung-up on documentation-only problems.  
(Of course I hope that every time someone adds a new feature they also update the documentation, in which case there's no time savings.)

**Comment by sidharthv96** [comment]
>  Just trying to get a handle on what the priorities / loads/ costs are for this project.

I don't think we have that defined anywhere (a good contender in contributin.md maybe?).
But my instinct is to use native functions if it's available and good enough for the job without adding too much complexity, and fallback to external deps if they're easier to integrate. 

https://github.com/mermaid-js/mermaid/pull/3434 is a good example. We were already using lodash as dependency, and the handwritten memoize function was causing us some trouble.
